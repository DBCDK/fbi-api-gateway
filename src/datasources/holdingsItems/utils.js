// Helper function to construct the URL dynamically
export function buildPath(
  url,
  { agencyId, bibliographicRecordId, trackingId, itemId }
) {
  const basePath = [url, agencyId, bibliographicRecordId, itemId]
    .filter(Boolean)
    .join("/");
  const query = trackingId
    ? `?trackingId=${encodeURIComponent(trackingId)}`
    : "";
  return `${basePath}${query}`;
}
