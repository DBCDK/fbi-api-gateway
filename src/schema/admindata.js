import { get } from "lodash";

export const typeDef = `
type AdminData {
  creationDate: String
  requestButton: Boolean
}`;

export const resolvers = {
  AdminData: {
    async creationDate(parent, args, context, info) {
      const manifestation = await context.datasources
        .getLoader("openformat")
        .load(parent.pid);
      return get(manifestation, "admindata.creationDate.$", "unknown");
    },
    async requestButton(parent, args, context, info) {
      const manifestation = await context.datasources
        .getLoader("openformat")
        .load(parent.pid);
      return get(manifestation, "admindata.requestButton.$") === "true";
    },
  },
};
