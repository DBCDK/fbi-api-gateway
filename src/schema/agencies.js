export const typeDef = `
type LatitudeLongitude {
  """
  Geolocation latitude
  """
  latitude: Float!
  
  """
  Geolocation longitude
  """
  longitude: Float!
}

type AgencyList {
  """
  Total number of agencies with branches that match query
  """
  agencyCount: Int!

  """
  List of agencyIds with branches that match query
  """
  agencyIds: [String!]!
  
  """
  Total number of branches that match query (across all resulting agencies)
  """
  branchCount: Int!
  
  """
  List of branchIds of branches that match query (across all resulting agencies)
  """
  branchIds: [String!]!

  """
  Agencies with branches that match query
  """
  agencies: [Agency!]!
}

type Agency {
  """
  Id of the agency (queryable)
  """
  agencyId: String!
  
  """
  Name of the agency (queryable)
  """
  agencyName: String!
  
  """
  Agency phonenumber (likely the main branch's number or null)
  """
  agencyPhone: String
  
  """
  Agency email (likely the main branch's number or null)
  """
  agencyEmail: String
  
  """
  Number of branches that match query within agency
  """
  branchCount: Int!
  
  """
  List of branchIds of branches that match query within agency
  """
  branchIds: [String!]!
  
  """
  Highlights specific to the agency
  """
  highlights: [Highlight!]!
  
  """
  Branches that match query within agency
  """
  agencyBranches: [AgencyBranch!]!
}

type AgencyBranch {
  """
  Id of the branch's agency (queryable)
  """
  agencyId: String!
  
  """
  Name of the branch's agency (queryable)
  """
  agencyName: String!

  """
  Id of the branch (queryable)
  """
  branchId: String!

  """
  Name of the branch (queryable)
  """
  branchName: [String!]!

  """
  Shortname of the branch (queryable)
  """
  branchShortName: [String!]!

  """
  Branch's phonenumber
  """
  branchPhone: String

  """
  Branch's email
  """
  branchEmail: String

  """
  Branch's InterLibraryLoan email
  """
  branchIllEmail: String

  """
  The city the branch is located in (queryable)
  """
  city: String

  """
  The geolocation of the branch
  """
  geolocation: LatitudeLongitude
  
  """
  Highlights specific to the branch and its agency
  """
  highlights: [Highlight!]!  

  """
  Opening hours of the branch (can be a html-element)
  """
  openingHours: [String!]!

  """
  Address of the branch (queryable)
  """
  postalAddress: String

  """
  PostalCode of the branch (queryable)
  """
  postalCode: Int
}
`;

function highlightResolver(parent, includedFields = null) {
  if (!parent.highlights) {
    return [];
  }

  return Object.entries(parent.highlights)
    .map(([key, value]) => ({
      key,
      value,
    }))
    .filter((highlight) => highlight.value.includes("<mark>"))
    .filter((highlight) =>
      includedFields === null ? true : includedFields.includes(highlight.key)
    );
}

export const resolvers = {
  Agency: {
    highlights(parent, args, context, info) {
      return highlightResolver(parent, ["agencyName", "agencyId"]);
    },
  },
  AgencyBranch: {
    highlights(parent, args, context, info) {
      return highlightResolver(parent);
    },
  },
};
