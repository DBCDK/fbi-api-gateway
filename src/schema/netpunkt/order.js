/**
 * @file Provides direct access for Netpunkt to Openorder.
 *
 * This functionality is necessary to allow librarians to create
 * loans on behalf of end users, ensuring efficient service
 * for library operations.
 */

export const typeDef = `

  input NetpunktSubmitOrderInput {
    orderType: OrderTypeEnum,
    pids: [String!]!,
    pickUpBranch: String!,
    key: String,
    exactEdition: Boolean
    """
    expires is required to be iso 8601 dateTime eg. "2024-03-15T12:24:32Z"
    """
    expires: String
    userParameters: SubmitOrderUserParametersInput!
    author: String
    authorOfComponent: String
    pagination: String
    publicationDate: String
    publicationDateOfComponent: String
    title: String
    titleOfComponent: String
    volume: String
    requesterInitials: String
  }
  type Netpunkt {
    """
    Submits an order from Netpunkt to OpenOrder on behalf of an end user
    """
    submitOrder(input: NetpunktSubmitOrderInput!, dryRun: Boolean): SubmitOrder
  }
  extend type Mutation {
    """
    Netpunkt related mutations
    """
    netpunkt: Netpunkt!
  }
`;

export const resolvers = {
  Mutation: {
    netpunkt() {
      return {};
    },
  },
  Netpunkt: {
    async submitOrder(parent, args, context) {
      if (args?.dryRun) {
        return { status: "OWNED_ACCEPTED", orderId: "123456" };
      }

      // The userId of the Netpunkt user (not the end user)
      const authUserId =
        context?.user?.userId || args?.input?.requesterInitials;

      const pickupBranch = args?.input?.pickUpBranch;

      const branch = (
        await context.datasources.getLoader("library").load({
          branchId: pickupBranch,
        })
      ).result?.[0];

      if (!branch) {
        return {
          status: "UNKNOWN_PICKUPAGENCY",
        };
      }

      return await context.datasources.getLoader("submitOrder").load({
        branch,
        input: args.input,
        accessToken: context.accessToken,
        smaug: context.smaug,
        authUserId,
      });
    },
  },
};
