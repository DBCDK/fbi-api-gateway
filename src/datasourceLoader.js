import DataLoader from "dataloader";

import { log } from "dbc-node-logger";
import { withRedis, clearRedis } from "./datasources/redis.datasource";
import { createFetchWithConcurrencyLimit } from "./utils/fetchWithLimit";
import { getFilesRecursive } from "./utils/utils";
import config from "./config";

/*
  This is how we time parallel requests to a single datasource.
  
  Request A
  ------------
          Request B
          ------------------
                                      Request C
                                      -----------
  Total AB
  --------------------------
  
  The totaltime for the requests should NOT be:
  Request A + Request B + Request C
  
  But instead:
  Total AB + Request C
*/

export const trackMe = () => {
  return {
    start: function (uuid) {
      this.uuid = uuid;
      return this;
    },
    createTracker: function (name) {
      // If this datasource has not been tracked yet, initialize its tracking object.
      if (!this.trackObject[name]) {
        this.trackObject[name] = {
          overlappingRequestCount: 0, // Number of ongoing requests for this datasource.
          count: 0, // Total number of elements retrieved from datasource
          started: 0, // Timestamp when the first concurrent request starts.
          time: 0, // Total time taken for all requests so far.
          sequentialTime: 0, // If requests were to be run sequentially, this is the total time
          batches: 0, // Dataloader batches multiple requests, into a single request when possible
        };
      }
    },

    // Function to indicate the beginning of a datasource call.
    begin: function (name, count) {
      this.createTracker(name);

      const beginTime = performance.now();

      const trackObject = this.trackObject[name];

      // Increment the count of ongoing requests.
      trackObject.overlappingRequestCount++;

      // Increment total number of elements retrieved from datasource
      trackObject.count += count;
      trackObject.batches++;

      // Function to indicate the end of an datasource call.
      return function end() {
        const endTime = performance.now();
        const duration = Math.round(endTime - beginTime);
        // Decrement the count of ongoing requests.
        trackObject.overlappingRequestCount--;

        // If all ongoing requests are complete, calculate the time taken.
        if (trackObject.overlappingRequestCount === 0) {
          trackObject.time += duration;
        }

        trackObject.sequentialTime += duration;
      };
    },

    cacheMiss: function (name, count) {
      this.createTracker(name);

      this.trackObject[name].cacheMiss =
        (this.trackObject[name].cacheMiss || 0) + count;
    },
    uuid: null,
    trackObject: {},
    getMetrics: function () {
      const res = {};

      Object.entries(this.trackObject).forEach(([key, val]) => {
        res[key] = {
          count: val.count,
          time: val.time,
          cacheMiss: val.cacheMiss,
          avg_ms: Math.round(val.sequentialTime / val.count),
        };
      });

      return res;
    },
  };
};

// Find all datasources in src/datasources
export const datasources = getFilesRecursive(`${__dirname}/datasources`)
  .map((file) => {
    if (!file.path.endsWith(".datasource.js")) {
      return;
    }
    const { load, options, batchLoader, status, testLoad } = require(file.path);
    if (!load && !batchLoader) {
      return;
    }

    // Extract datasource name from filename
    const name = file.file.replace(".datasource.js", "");

    return {
      testLoad,
      load,
      batchLoader,
      name,
      options,
      statusChecker: status ? () => status(load) : null,
    };
  })
  .filter((func) => !!func);

const nameToDatasource = {};
datasources.forEach((ds) => (nameToDatasource[ds.name] = ds));

log.debug(
  `found ${datasources.length} datasources, ${datasources
    .map((datasource) => datasource.name)
    .join(", ")}`
);

/**
 * This will initialize a single dataloader.
 * It makes sure a context (with a concurrency limited fetcher), is injected
 * into the batchloader function.
 *
 * It sets up Redis for the dataloader
 *
 * It collects performance metrics for the datasource.
 */
function setupDataloader(
  { name, load, options, batchLoader, testLoad },
  context
) {
  const testModeActivated = context?.testUser && testLoad;
  const loadImpl = testModeActivated ? testLoad : load;

  let batchLoaderWithContext = batchLoader
    ? (keys) => batchLoader(keys, context)
    : (keys) => Promise.all(keys.map((key) => loadImpl(key, context)));

  if (!testModeActivated && options?.redis?.prefix && options?.redis?.ttl) {
    batchLoaderWithContext = withRedis(batchLoaderWithContext, {
      ...options.redis,
      ...context,
      datasourceName: name,
    });
  }

  async function batchLoaderWithTiming(keys) {
    const end = context?.track?.begin?.(name, keys.length);

    try {
      return await batchLoaderWithContext(keys);
    } catch (err) {
      // Wraps the load call in a try catch, in order to capture the
      // stack trace before dataloader is called. Otherwise we would
      // be limited to see the stack trace from the data sources, but
      // not the GraphQL resolvers.
      // https://brunoscheufler.com/blog/2021-07-25-improving-dataloader-stack-traces
      if (err instanceof Error && err.stack) {
        const currentStack = { stack: "" };
        Error.captureStackTrace(currentStack, batchLoaderWithTiming);
        err.stack += currentStack.stack.replace("Error", "\nBefore DataLoader");
      }
      throw err;
    } finally {
      end?.();
    }
  }

  const loader = new DataLoader(batchLoaderWithTiming, {
    // If key is an object, we stringify
    // to make it useful as a cache key
    cacheKeyFn: (key) => (typeof key === "object" ? JSON.stringify(key) : key),
    maxBatchSize: 100,
  });

  if (options?.redis?.prefix) {
    loader.clearRedis = (key) => clearRedis(options?.redis?.prefix, key);
  }

  return {
    loader,
  };
}

/**
 * Will instantiate dataloaders from datasources.
 * This should be done for every incoming GraphQL request
 */
export default function createDataLoaders(uuid, testUser, accessToken) {
  const result = {};
  const track = trackMe().start(uuid);

  result.trackingObject = track;
  const fetchWithConcurrencyLimit = createFetchWithConcurrencyLimit(
    config.fetchConcurrencyLimit
  );
  // Gets a loader by name.
  // A loader will be initialized first time it is called
  function getLoader(name) {
    if (!result[name]) {
      result[name] = setupDataloader(nameToDatasource[name], {
        getLoader,
        track,
        fetch: (url, options) => fetchWithConcurrencyLimit(url, options, name),
        trackingId: uuid,
        testUser,
        accessToken,
      })?.loader;
    }
    return result[name];
  }
  result.getLoader = getLoader;
  return result;
}

/**
 * For testing
 */
export function createMockedDataLoaders() {
  const mockedDatasources = {};
  getFilesRecursive("./src/datasources/mocked")
    .filter(
      (file) =>
        file.path.endsWith("datasource.mocked.js") && require(file.path).load
    )
    .map((file) => ({
      ...file,
      name: file.file.replace(".datasource.mocked.js", ""),
      load: require(file.path).load,
      clearRedis: () => {},
    }))
    .forEach((loader) => {
      mockedDatasources[loader.name] = loader;
    });

  mockedDatasources.getLoader = (name) => mockedDatasources[name];

  return mockedDatasources;
}

export function createTestUserDataLoaders() {
  const mockedDataLoaders = createMockedDataLoaders();
  getFilesRecursive("./src/datasources")
    .filter(
      (file) =>
        file.path.endsWith("datasource.js") && require(file.path).testLoad
    )
    .map((file) => ({
      ...file,
      name: file.file.replace(".datasource.js", ""),
      load: require(file.path).testLoad,
      clearRedis: () => {},
    }))
    .forEach((loader) => {
      mockedDataLoaders[loader.name] = loader;
    });

  return mockedDataLoaders;
}
