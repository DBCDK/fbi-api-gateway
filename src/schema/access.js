import { resolveAccess } from "../utils/access";

export const typeDef = `
enum AccessTypeCodeEnum {
  PHYSICAL
  ONLINE
  UNKNOWN
}
enum AccessUrlTypeEnum {
  IMAGE
  OTHER
  RESOURCE
  SAMPLE
  TABLE_OF_CONTENTS
  THUMBNAIL
}

enum LinkStatusEnum {
  BROKEN
  GONE
  INVALID
  OK
}

type AccessType {
  display: String!
  code: AccessTypeCodeEnum!
}
type Ereol {
  """
  The origin, e.g. "Ereolen" or "Ereolen Go"
  """
  origin: String!

      """
  The url where manifestation is located
  """
  url: String!

  """
  Is this a manifestation that always can be loaned on ereolen.dk even if you've run out of loans this month
  """
  canAlwaysBeLoaned: Boolean!

  """
  Notes for the resource
  """
  note: String
}
type AccessUrl {
  """
  The origin, e.g. "DBC Webarkiv"
  """
  origin: String!
  
  """
  Description/type of URL
  """
  urlText: String

  """
  The url where manifestation is located
  """
  url: String!
  
  """
  Notes for the resource
  """
  note: String
  
  """
  If the resource requires login
  """
  loginRequired: Boolean!

  """
  The type of content that can be found at this URL
  """
  type: AccessUrlTypeEnum
  
  """
  Status from linkcheck
  """
  status: LinkStatusEnum!
}
type InterLibraryLoan {
  """
  Is true when manifestation can be borrowed via ill
  """
  loanIsPossible: Boolean!
  """
  Is newly added - nice to know if there are no localizations
  """
  accessNew: Boolean!  
}
type InfomediaService {
  """
  Infomedia ID which can be used to fetch article through Infomedia Service
  """
  id: String!
}
type DigitalArticleService {
  """
  Issn which can be used to order article through Digital Article Service
  """
  issn: String!
}

union AccessUnion = AccessUrl | Ereol | InterLibraryLoan | InfomediaService | DigitalArticleService
`;

export const resolvers = {
  Manifestation: {
    async access(parent, args, context, info) {
      return resolveAccess(parent, context);
    },
  },
};
