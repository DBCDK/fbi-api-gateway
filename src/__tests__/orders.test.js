/**
 * @file This file contains tests for submitMutliOrder
 *
 */

import { createMockedDataLoaders } from "../datasourceLoader";
import { performTestQuery } from "../utils/utils";
import { getAvailablePids } from "../schema/order";

const userParameters = {
  userName: "Test Testesen",
  userMail: "test@dbc.dk",
};

const query = `
          mutation Example_SubmitMultipleOrders($input: SubmitMultipleOrdersInput!) {
            submitMultipleOrders(input: $input, dryRun: false) {
              failedAtCreation
              successfullyCreated
              status
              ok
            }
          }`;

const context = {
  datasources: createMockedDataLoaders(),
  accessToken: "DUMMY_TOKEN",
  user: {
    userId: "some-id",
    loggedInAgencyId: "715100",
    municipalityAgencyId: "100200", //TODO test users dont have digital article service, so its always a physical order
  },
  smaug: {
    app: { id: "app-name", ips: ["1.1.1.1"] },
    orderSystem: "bibliotekdk_21",
  },
};

const contextNotSubscriberOfDigitalArticleService = {
  ...context,
  user: {
    userId: "some-id",
    loggedInAgencyId: "715100",
    municipalityAgencyId: "911116", //911116 is not subscriber of digital article service as per 8. dec. 2023
  },
};
const contextNoAccountAtMunicipality = {
  ...context,
  user: {
    userId: "some-id",
    loggedInAgencyId: "715100",
    municipalityAgencyId: "734000", // DUMMY_TOKEN is not user at 734000
  },
};

const book = {
  pids: ["870970-basis:62795522", "870970-basis:39074184"],
  key: "work-of:870970-basis:39074184Bog",
};

const book2 = {
  pids: ["870970-basis:134976977", "870970-basis:62371455"],
  key: "work-of:870970-basis:62371455Bog",
};

const invalidOrder = {
  pids: [],
  key: "work-of:idontexistBog",
};

const digitalArticle = {
  pids: ["870970-basis:04978617"],
  key: "work-of:870970-basis:04978617Tidsskrift",
  periodicaForm: {
    publicationDateOfComponent: "2023",
    titleOfComponent: "Some exciting article",
    pid: "870970-basis:04978617",
  },
};

const physicalPeriodica = {
  pids: ["870970-basis:04978617"],
  key: "work-of:870970-basis:04978617Tidsskrift",
  periodicaForm: {
    publicationDateOfComponent: "2023",
    pid: "870970-basis:04978617",
  },
};

describe("submitMultipleOrders", () => {
  test("succeeds: user orders multiple books ", async () => {
    const result = await performTestQuery({
      query,
      variables: {
        input: {
          materialsToOrder: [book, book2],
          pickUpBranch: "715100",
          userParameters,
        },
      },
      context,
    });
    expect(result).toMatchSnapshot();
  });

  test("succeeds: user orders multiple books and digital article", async () => {
    const result = await performTestQuery({
      query,
      variables: {
        input: {
          materialsToOrder: [book, book2, digitalArticle],
          pickUpBranch: "715100",
          userParameters,
        },
      },
      context,
    });
    expect(result).toMatchSnapshot();
  });

  test("succeeds: user orders a book, a digital article and a physical periodica", async () => {
    const result = await performTestQuery({
      query,
      variables: {
        input: {
          materialsToOrder: [book, digitalArticle, physicalPeriodica],
          pickUpBranch: "715100",
          userParameters,
        },
      },
      context,
    });
    expect(result).toMatchSnapshot();
  });

  test("succeeds: digital article becomes physical order, if users homeagency not subscriber of digital article service", async () => {
    const result = await performTestQuery({
      query,
      variables: {
        input: {
          materialsToOrder: [book, digitalArticle],
          pickUpBranch: "715100",
          userParameters,
        },
      },
      context: contextNotSubscriberOfDigitalArticleService,
    });
    expect(result).toMatchSnapshot();
  });

  test("succeeds: digital article becomes physical order, if user doesnt have account at homeagency", async () => {
    const result = await performTestQuery({
      query,
      variables: {
        input: {
          materialsToOrder: [book, digitalArticle],
          pickUpBranch: "715100",
          userParameters,
        },
      },
      context: contextNoAccountAtMunicipality,
    });
    expect(result).toMatchSnapshot();
  });

  test("all orders fail: unknown pickupBranch", async () => {
    const result = await performTestQuery({
      query,
      variables: {
        input: {
          materialsToOrder: [book, digitalArticle],
          pickUpBranch: "123", //submitOrder fails with UNKNOWN_PICKUPAGENCY
          userParameters,
        },
      },
      context,
    });
    expect(result).toMatchSnapshot();
  });

  test("1 succeeds 1 fails", async () => {
    const result = await performTestQuery({
      query,
      variables: {
        input: {
          materialsToOrder: [book, invalidOrder],
          pickUpBranch: "715100",
          userParameters,
        },
      },
      context,
    });
    expect(result).toMatchSnapshot();
  });

  test("2 succeed, we order physical periodica when no title, author or pagenumber is given", async () => {
    const result = await performTestQuery({
      query,
      variables: {
        input: {
          materialsToOrder: [book, physicalPeriodica],
          pickUpBranch: "715100",
          userParameters,
        },
      },
      context,
    });
    expect(result).toMatchSnapshot();
  });

  test("getAvailablePids filters PIDs based on availability", async () => {
    // Mock the holdingsGetAllAvailability datasource
    const mockHoldingsLoader = {
      load: jest.fn().mockImplementation(({ pid }) => {
        // Mock that some PIDs are not available (librariesLend = 0)
        const availabilityMap = {
          "870970-basis:62795522": { librariesLend: 2 }, // Available
          "870970-basis:39074184": { librariesLend: 0 }, // Not available
          "870970-basis:134976977": { librariesLend: 1 }, // Available
          "870970-basis:62371455": { librariesLend: 0 }, // Not available
        };
        return Promise.resolve(availabilityMap[pid] || { librariesLend: 0 });
      }),
    };

    const testContext = {
      datasources: {
        getLoader: jest.fn().mockReturnValue(mockHoldingsLoader),
      },
      smaug: {
        gateway: {
          localizationsRole: "test-role",
        },
      },
    };

    const pids = [
      "870970-basis:62795522",
      "870970-basis:39074184",
      "870970-basis:134976977",
      "870970-basis:62371455",
    ];
    const result = await getAvailablePids(pids, testContext);

    // Should return only available PIDs
    expect(result).toHaveLength(2);
    expect(result[0]).toBe("870970-basis:62795522");
    expect(result[1]).toBe("870970-basis:134976977");

    // Verify that holdingsGetAllAvailability was called for each PID
    expect(mockHoldingsLoader.load).toHaveBeenCalledTimes(4);
    expect(mockHoldingsLoader.load).toHaveBeenCalledWith({
      pid: "870970-basis:62795522",
      role: "test-role",
    });
    expect(mockHoldingsLoader.load).toHaveBeenCalledWith({
      pid: "870970-basis:39074184",
      role: "test-role",
    });
    expect(mockHoldingsLoader.load).toHaveBeenCalledWith({
      pid: "870970-basis:134976977",
      role: "test-role",
    });
    expect(mockHoldingsLoader.load).toHaveBeenCalledWith({
      pid: "870970-basis:62371455",
      role: "test-role",
    });
  });
});
