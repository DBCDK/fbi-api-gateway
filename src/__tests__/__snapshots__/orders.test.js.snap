// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`submitMultipleOrders 1 succeeds 1 fails 1`] = `
Object {
  "data": Object {
    "submitMultipleOrders": Object {
      "failedAtCreation": Array [
        "work-of:idontexistBog",
      ],
      "ok": false,
      "status": "OK",
      "successfullyCreated": Array [
        "work-of:870970-basis:39074184Bog",
      ],
    },
  },
}
`;

exports[`submitMultipleOrders all orders fail: unknown pickupBranch 1`] = `
Object {
  "data": Object {
    "submitMultipleOrders": Object {
      "failedAtCreation": Array [
        "work-of:870970-basis:39074184Bog",
        "work-of:870970-basis:04978617Tidsskrift",
      ],
      "ok": false,
      "status": "UNKNOWN_PICKUPAGENCY",
      "successfullyCreated": Array [],
    },
  },
}
`;

exports[`submitMultipleOrders succeeds: digital article becomes physical order, if user doesnt have account at homeagency 1`] = `
Object {
  "data": Object {
    "submitMultipleOrders": Object {
      "failedAtCreation": Array [],
      "ok": true,
      "status": "OK",
      "successfullyCreated": Array [
        "work-of:870970-basis:39074184Bog",
        "work-of:870970-basis:04978617Tidsskrift",
      ],
    },
  },
}
`;

exports[`submitMultipleOrders succeeds: digital article becomes physical order, if users homeagency not subscriber of digital article service 1`] = `
Object {
  "data": Object {
    "submitMultipleOrders": Object {
      "failedAtCreation": Array [],
      "ok": true,
      "status": "OK",
      "successfullyCreated": Array [
        "work-of:870970-basis:39074184Bog",
        "work-of:870970-basis:04978617Tidsskrift",
      ],
    },
  },
}
`;

exports[`submitMultipleOrders succeeds: user orders a book, a digital article and a physical article 1`] = `
Object {
  "data": Object {
    "submitMultipleOrders": Object {
      "failedAtCreation": Array [],
      "ok": true,
      "status": "OK",
      "successfullyCreated": Array [
        "work-of:870970-basis:39074184Bog",
        "work-of:870970-basis:04978617Tidsskrift",
        "work-of:870970-basis:04978617Tidsskrift",
      ],
    },
  },
}
`;

exports[`submitMultipleOrders succeeds: user orders multiple books  1`] = `
Object {
  "data": Object {
    "submitMultipleOrders": Object {
      "failedAtCreation": Array [],
      "ok": true,
      "status": "OK",
      "successfullyCreated": Array [
        "work-of:870970-basis:39074184Bog",
        "work-of:870970-basis:62371455Bog",
      ],
    },
  },
}
`;

exports[`submitMultipleOrders succeeds: user orders multiple books and digital article 1`] = `
Object {
  "data": Object {
    "submitMultipleOrders": Object {
      "failedAtCreation": Array [],
      "ok": true,
      "status": "OK",
      "successfullyCreated": Array [
        "work-of:870970-basis:39074184Bog",
        "work-of:870970-basis:62371455Bog",
        "work-of:870970-basis:04978617Tidsskrift",
      ],
    },
  },
}
`;
